cirq-core/cirq/ops/conditional_operation.py

```diff
@@ -0,0 +1,161 @@
+# Copyright 2021 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import (
+    AbstractSet,
+    Any,
+    cast,
+    Dict,
+    Optional,
+    Sequence,
+    TYPE_CHECKING,
+    Tuple,
+    Union,
+)
+
+from cirq import protocols, value
+from cirq.ops import raw_types
+
+if TYPE_CHECKING:
+    import cirq
+
+
+@value.value_equality
+class ConditionalOperation(raw_types.Operation):
+    """Augments existing operations to be conditionally executed."""
+
+    def __init__(
+        self,
+        sub_operation: 'cirq.Operation',
+        controls: Sequence[Union[str, 'cirq.MeasurementKey']],
+    ):
+        controls = tuple(value.MeasurementKey(k) if isinstance(k, str) else k for k in controls)
+        if isinstance(sub_operation, ConditionalOperation):
+            # Auto-flatten nested controlled operations.
```

Comment:
This idea was copied from ControlledOperaton and TaggedOperation. But I'm not so sure it's the right thing to do here. For instance, consider `ConditionalOp(TaggedOp(ConditionalOp))`. Should this expose the inner controls or only the outer? How about `ConditionalOp(TaggedOp(Subcircuit(ConditionalOp)))`? These should be exposed in the `control_keys` protocol I assume, but do they need to be in `op.controls`? Or should `op.controls` only show the top level controls? I could see this either way.